src/coreclr/tools/dotnet-pgo/MibcEmitter.cs

```diff
@@ -0,0 +1,330 @@
+                    }
+                    if ((processedMethodData.CallWeights != null) && processedMethodData.CallWeights.Count > 0)
+                    {
+                        _il.LoadString(_emitter.GetUserStringHandle("WeightedCallData"));
+                        _il.LoadConstantI4(processedMethodData.CallWeights.Count);
+                        foreach (var entry in processedMethodData.CallWeights)
+                        {
+                            EntityHandle calledMethod = _emitter.GetMethodRef(entry.Key);
+                            _il.OpCode(ILOpCode.Ldtoken);
+                            _il.Token(calledMethod);
+                            _il.LoadConstantI4(entry.Value);
+                        }
+                    }
+                    if (processedMethodData.SchemaData != null)
+                    {
+                        _il.LoadString(_emitter.GetUserStringHandle("InstrumentationDataStart"));
+                        PgoProcessor.EncodePgoData<TypeSystemEntityOrUnknown>(processedMethodData.SchemaData, this, true);
+                    }
+                    _il.OpCode(ILOpCode.Pop);
+                }
+                catch (Exception ex)
+                {
+                    Program.PrintWarning($"Exception {ex} while attempting to generate method lists");
+                }
+            }
+
+            public MethodDefinitionHandle EmitMethod()
+            {
+                s_emitCount++;
+                string basicName = "Assemblies_" + _name;
+                if (_name.Length > 200)
+                    basicName = basicName.Substring(0, 200); // Cap length of name at 200, which is reasonably small.
+
+                string methodName = basicName + "_" + s_emitCount.ToString(CultureInfo.InvariantCulture);
+                return _emitter.AddGlobalMethod(methodName, _il, 8);
+            }
+
+            bool IPgoEncodedValueEmitter<TypeSystemEntityOrUnknown>.EmitDone()
+            {
+                _il.LoadString(_emitter.GetUserStringHandle("InstrumentationDataEnd"));
+                return true;
+            }
+
+            void IPgoEncodedValueEmitter<TypeSystemEntityOrUnknown>.EmitLong(long value, long previousValue)
+            {
+                if ((value <= int.MaxValue) && (value >= int.MinValue))
+                {
+                    _il.LoadConstantI4(checked((int)value));
+                }
+                else
+                {
+                    _il.LoadConstantI8(value);
+                }
+            }
+
+            void IPgoEncodedValueEmitter<TypeSystemEntityOrUnknown>.EmitType(TypeSystemEntityOrUnknown type, TypeSystemEntityOrUnknown previousValue)
+            {
+                if (type.AsType != null)
+                {
+                    _il.OpCode(ILOpCode.Ldtoken);
+                    _il.Token(_emitter.GetTypeRef(type.AsType));
+                }
+                else
+                    _il.LoadConstantI4(type.AsUnknown);
```

Comment:
See MibcProfileParser.IPgoSchemaDataLoader<TypeSystemEntityOrUnknown>.TypeFromLong. I've tweaked the format though, as relying on a catch is a poor choice. It will now reliably have zeroes in the upper byte.

Reply:
I see, thanks for clarifying.